### ISSUE TITLE
Wireguard Forgets Apps 0.5.5e

### ISSUE THREAD

**Issue by luckygitt** [celzero/rethink-app#1370]
I just noticed after some preiod of time, that my Wireguard config is forgetting (ie. unchecking) included apps.

For your info, I have 294 apps, most of them blocked, but all checked to go through one wireguard server. After some period of time, when I look, only 287 are now included. Most notable one that was now unchecked is Google Play Services - a quck look suggested all the others were blocked anyway.

This wasn't a "one off", I've noticed this several times already today.

Edit: Lockdown - enabled, Always-On - disabled

**Comment by ignoramous** [comment]
Thanks. This sounds worrying and unusual.

Every 3 hours or so, apps list is refreshed. If there's a bug in there, it could be that these apps are auto-removed. Though, that should only happen if the app was uninstalled or reinstalled; not otherwise.

If this happens frequently, can you capture `adb logcat` and share it with us (`mz` at `celzero` dot `com` and mention this issue)?

1. Add all these apps you want to.
2. Tap on the "refresh" icon in the search bar in *Configure* -> *Apps* UI.
3. See if apps are removed from your WireGuard profile.
4. If so, stop `adb logcat` and email the output, if you're comfortable doing so.

**Comment by luckygitt** [comment]
Email sent to "MZ @"

**Comment by ignoramous** [comment]
Thanks. Did you tap on `Refresh` when you took the log?

Believe this could be it. It looks like this happened when the app was stop and started again:

```
I AppDatabase: sizes: rmv: 0; add: 0; update: 0
04-16 08:12:45.593 14121 14230 W ProxyLogs: delete dup pxms: [AppInfoTuple(uid=1002, packageName=com.xiaomi.bluetooth), AppInfoTuple(uid=1073, packageName=com.google.android.cellbroadcastservice), AppInfoTuple(uid=10176, packageName=com.google.android.gms), AppInfoTuple(uid=10056, packageName=com.android.providers.userdictionary), AppInfoTuple(uid=10070, packageName=com.android.mtp), AppInfoTuple(uid=1000, packageName=com.miui.screenrecorder), AppInfoTuple(uid=1001, packageName=com.mediatek.telephony)]
```

**Comment by luckygitt** [comment]
> Did you tap on `Refresh` when you took the log?

No, as this didnt make any difference / solve the problem.

> Believe this could be it. It looks like this happened when the app was stop and started again:

These are indeed the apps that become unchecked / leak.

The stop/start isn't the main cause of the gltch, as this happenes on its own after an undetermined period of time without any external  interfence (ie. just not using the phone for some time).

**Comment by ignoramous** [comment]
Thanks.

A subtle database cache related bug in the [ProxyManager code](https://github.com/celzero/rethink-app/blob/2987824b0a1e3822b262d5908db494ef2a218f5a/app/src/full/java/com/celzero/bravedns/service/ProxyManager.kt) exists, though we can't for sure root-caused it to the issue you're seeing but are fairly certain it causes it.

We're working on a fix which will be part of `v055f` (due this week).

**Comment by ignoramous** [comment]
[`v055f`](https://github.com/celzero/rethink-app/releases/tag/v0.5.5f) is a go. F-Droid and Play Store should get the update within a week. Please test it and let us know if this issue is fixed. Thanks.

**Comment by ignoramous** [comment]
> After one or two attempts the app crashes (a new bug?)

Is this crash 100% reproducible? If so, simply stopping / starting enough?

> In fact it is worse

You mean, more apps are unselected or just as before? I'm not able to reproduce this issue on my Android, but we'll keep looking if there's something else (apart from what we did for `v055f`) we are missing to fix this issue.

**Comment by luckygitt** [comment]
> > After one or two attempts the app crashes (a new bug?)
> 
> Is this crash 100% reproducible? If so, simply stopping / starting enough?

Yes, stop is enough, sometimes it takes 2 or 3 stops/starts for app to crash - when it does, the same apps are unchecked in Wireguard.


> > In fact it is worse
> 
> You mean, more apps are unselected or just as before? I'm not able to reproduce this issue on my Android, but we'll keep looking if there's something else (apart from what we did for `v055f`) we are missing to fix this issue.

No, the same apps affected but the "glitch" seems more easily reproducable (for example, the start/stop mentioned above.

**Comment by ignoramous** [comment]
> Yes, stop is enough, sometimes it takes 2 or 3 stops/starts for app to crash - when it does, the same apps are unchecked in Wireguard.

Okay, I see a crash but not sure if it the same one you're seeing; so, if you're technical enough to grab logs and verify, please do so: #1386

> No, the same apps affected but the "glitch" seems more easily reproducable (for example, the stop/start mentioned above).

Worrying.

**Comment by luckygitt** [comment]
Log and screenshots sent to "MZ@".

**Comment by luckygitt** [comment]
Initial testing seems to confirm that ths Forgetting App Problem ("multi package apps auto-deleting from WireGuard configs") was fixed in 5.5h.
I will report back if further testing shows otherwise.

However, I can still manage to crash the app if I stop and start it a few times (ie. not fixed).

**Comment by ignoramous** [comment]
Thanks for confirming this current is no longer exists. Also, thanks for your patience and logs (without which we'd not have been able to fix this issue at all).

As for the crash you sent us over email, that has *not* been fixed in `v055h` but we will look to close it in `v055i`:
- https://github.com/celzero/rethink-app/issues/1395
